### PR TITLE
Support constraints with Lowpass[Cell] and Alpha[Cell]

### PR DESCRIPTION
**Pull request by arvoelke**
Addresses #15 by taking the suggested approach of supporting constraints, and implementing a constraint that sets all of the time constants to their mean (and optionally making that non-negative).

TODO:
 - [x] After collecting feedback on this implementation, the solution could be replicated across `Lowpass` and all of the other filter cells and RNN layers.
 - [x] Update changelog

### CODE REVIEW

File: keras_spiking/constraints.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+    def __init__(self, axis=-1, non_neg=False):
+        self.axis = axis
+        self.non_neg = non_neg
+        self._post_constraint = tf.keras.constraints.NonNeg() if non_neg else None
+
+    def __call__(self, w):
+        r = tf.repeat(
+            input=tf.math.reduce_mean(w, axis=self.axis, keepdims=True),
+            repeats=tf.shape(w)[self.axis],
```
Comment (hunse) [inline keras_spiking/constraints.py:23]: Do you know if this works if the shape for that dimension is `None`?

Comment (arvoelke) [reply keras_spiking/constraints.py:27]: I have not tried.

**Comment by arvoelke** [comment]
Note `keras_spiking/tests/test_layers.py::test_save_load[False]` is currently failing (while `use_cell=True` passes) I believe because the support has not yet been mirrored over from `LowpassCell` to `Lowpass`.

File: keras_spiking/constraints.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+            axis=self.axis,
+        )
+        if self._post_constraint is not None:
+            r = self._post_constraint(r)
+        return r
+
+    def get_config(self):
+        """Return config of layer (for serialization during model saving/loading)."""
+        return {"axis": self.axis, "non_neg": self.non_neg}
```
Comment (hunse) [inline keras_spiking/constraints.py:36]: This line is uncovered. Should probably add a `test_constraints.py` file that tests this, as well as a basic test of the constraint on e.g. a dense layer (which can be pretty similar to the one you added, it'd just be nice to have a unit test that's independent of the application to `LowpassCell`).

Comment (arvoelke) [reply keras_spiking/constraints.py:36]: I believe it will be convered when that failed test is made to pass (see comment in main PR).

Comment (arvoelke) [reply keras_spiking/constraints.py:36]: Wait sorry scratch that. I would need to actually use the constraint in that test to cover this line.

File: keras_spiking/constraints.py (hunk 1 above)
Comment (arvoelke) [reply keras_spiking/constraints.py:27]: I'm trying to think of a way to test this, but I'm not sure how to get into a situation where a weight tensor has a `None` dimension? Since this is always going to be passed a weight tensor via `Layer.add_weight` is it possible for this to happen?

File: keras_spiking/constraints.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+class Mean(tf.keras.constraints.Constraint):
+    """Constrains weight tensors to be their mean.
+
+    Parameters
+    ----------
+    axis : int
+        Axis used to compute the mean and repeat its value. Defaults to the last axis.
+    non_neg : boolean
+        Whether to constrain the resulting mean to be non-negative as well.
```
Comment (drasmuss) [inline keras_spiking/constraints.py:16]: We should probably just build this into the Lowpass layer (i.e., pass `tau` though a `relu` nonlinearity). Otherwise it'd be easy for people to accidentally shoot themselves in the foot by changing `tau_constraint`, and not realizing that they've lost the important non-negative property.

File: keras_spiking/layers.py [hunk 4]
```diff
@@ -535,6 +535,11 @@ class LowpassCell(KerasSpikingCell):
         then we often don't need to filter them either).
     level_initializer : str or ``tf.keras.initializers.Initializer``
         Initializer for filter state.
+    initial_level_constraint : str or ``tf.keras.constraints.Constraint``
+        Constraint for ``initial_level``.
+    tau_var_constraint : str or ``tf.keras.constraints.Constraint``
```
Comment (drasmuss) [inline keras_spiking/layers.py:540]: I'd just call this `tau_constraint` (to align with the other parameter in init here, which is just `tau`).

Comment (arvoelke) [reply keras_spiking/layers.py:540]: I was following this pattern: https://github.com/tensorflow/tensorflow/blob/v2.5.0/tensorflow/python/keras/layers/core.py#L1191-L1198 where it seems to be common for it to be named after the weights attribute that it belongs to (which is `tau_var` and not `tau`).

File: keras_spiking/constraints.py (hunk 3 above)
Comment (arvoelke) [reply keras_spiking/constraints.py:16]: I think this would be a breaking change potentially as gradient descent could behave differently if the clipping is happening inside of the inference call on an intermediate value rather than directly on the trained parameter after each gradient update. In the former case the underlying value of `tau_var` would be allowed to visit arbitrarily small negative values and potentially take a while to come back to positive, while the latter case it would always stay nonnegative and immediately be able to return to positive values if the gradient changes direction. So if this is the route being taken (clip tau in the inference call) I'd expect it to be a separate PR with a major version bump, and some checks to make sure it doesn't make it harder to train any important models/examples.

File: keras_spiking/layers.py (hunk 4 above)
Comment (hunse) [reply keras_spiking/layers.py:540]: Users don't see `tau_var`, though, unless they're really digging in to the implementation. I agree with Dan that just using `tau_constraint` is better.

**Comment by arvoelke** [comment]
Addressed the early feedback and TODO items. There is still this https://github.com/nengo/keras-spiking/pull/21#discussion_r674919732 that I'm not sure how to resolve. Tests are passing and coverage is at 100%, so this is ready for review.

File: keras_spiking/constraints.py (hunk 1 above)
Comment (hunse) [reply keras_spiking/constraints.py:27]: Yeah, so I guess the first question is does `Lowpass` even support having `None` for the size. From what I can tell from this simple script, it doesn't. This is actually something we might want a better error for, because right now it's a bit hard to parse (the error results in another error; the first error hints at the problem, but the second really obscures things). 
```
inp = x = tf.keras.Input([None, None])
x = keras_spiking.Lowpass(tau=0.005)(x)
model = tf.keras.Model(inp, [x])
y = model.predict(np.ones(1, 20, 5))
```
So in short, don't worry about supporting `None` here!

File: keras_spiking/layers.py (hunk 4 above)
Comment (arvoelke) [reply keras_spiking/layers.py:540]: Users will see the name `tau_var` if they are using TensorBoard to inspect the weights, or doing something like `[weight.name for weight in layer]`. In this case they will see `[initial_level, tau_var]` on one hand but then `[initial_level_constraint, tau_constraint]` on the other hand. So I'm not sure I see one as better than the other necessarily. But if there's no ideal solution then yeah I'm fine with either. Just explaining why I went this way rather than the other.

File: keras_spiking/constraints.py (hunk 1 above)
Comment (arvoelke) [reply keras_spiking/constraints.py:27]: I also tried the same thing with `Dense` and got the error: `ValueError: The last dimension of the inputs to `Dense` should be defined. Found `None`.` so this problem might not be unique to lowpass either. But it sounds like this could have a safety check to fail faster with a better error message (for dense it happens inside of `build` as early as possible).

File: keras_spiking/constraints.py (hunk 3 above)
Comment (arvoelke) [reply keras_spiking/constraints.py:16]: Thinking about this more and we could always do both in the mean time (keep constraining it as we are, and then additionally clip it as an extra safety precaution)? Might appear a little confusing/redundant though.

File: keras_spiking/layers.py (hunk 4 above)
Comment (drasmuss) [reply keras_spiking/layers.py:540]: We could rename `tau` to `tau_initializer` and then `tau_var` to `tau`. That'd be more consistent with Keras conventions (the current scheme was chosen to be more consistent with Nengo conventions, but we're trying to be closer to Keras in this repo).

File: keras_spiking/constraints.py (hunk 3 above)
Comment (drasmuss) [reply keras_spiking/constraints.py:16]: I suspect the difference between the two approaches would be minor. The underlying value of `tau_var` could drift slightly into the negative due to momentum terms in the optimizer, but I wouldn't expect that to be a major effect (that being said, I've never explicitly looked into it). We could also use something like `softplus` instead of `relu`, which would provide a more gradual thresholding. That would also be a breaking change as it'd change the effective value of `tau_var` in existing models, but we're in alpha so I'm not too worried about introducing breaking changes (as long as they're clearly marked by version changes and changelogs).

File: keras_spiking/layers.py (hunk 4 above)
Comment (arvoelke) [reply keras_spiking/layers.py:540]: Would the rename affect backwards compatibility of saved models/weights? If not this sounds like a good idea to me as it avoids this naming difference between `tau` and `tau_var` altogether. The main drawback is users would need to update how they've been calling the layer from `tau=` to `tau_intiializer=`. But if that's all ok would this be a separate PR?

File: keras_spiking/constraints.py (hunk 3 above)
Comment (arvoelke) [reply keras_spiking/constraints.py:16]: I looked into this a bit and I think you are right that the difference is basically not perceptable. I was leaning on some experience I had with a similar situation when training `1 / theta` in the LMU (I was seeing the underlying value become negative, even with clipping) but I think that was different enough that the effect doesn't apply here. I'll add that change to this PR.

File: keras_spiking/layers.py (hunk 4 above)
Comment (drasmuss) [reply keras_spiking/layers.py:540]: I'm fairly certain that the default format (SavedModel) will be fine as long as the order of the variables matches, which it will. I think the HDF5 format is a bit more fragile, so if someone has explicitly opted into that they might have problems, but I think that's a small enough disruption (especially considering we're in alpha, where breaking changes are expected).  

The `tau` -> `tau_initializer` change is a bit more drastic. But e could deprecate it more gradually (keeping around `tau`, which would set `tau_initializer=tf.initializers.constant(tau)`). But again given that we're in alpha I think it'd also be fine to go straight to the breaking change. I don't have a strong preference either way.

File: keras_spiking/constraints.py (hunk 1 above)
Comment (drasmuss) [reply keras_spiking/constraints.py:23]: `tf.shape(w)` will resolve the dynamic shape of `w` at runtime, so it would never be `None` (even if you could create a weight with an unspecified shape at build time, which as Aaron says I don't think is possible). So this should be fine :+1:

File: keras_spiking/layers.py (hunk 4 above)
Comment (drasmuss) [reply keras_spiking/layers.py]: Added a commit with this renaming. I went with kind of a middle ground. I didn't keep a deprecated ``tau`` parameter around, but I did make it so that you can pass in floats. So if people were using it as a positional argument, like `Lowpass(0.1)` , that will continue to work with no modifications needed.

**Review by drasmuss — APPROVED** [review]

File: keras_spiking/layers.py [hunk 5]
```diff
@@ -846,7 +906,7 @@ def call_inference(self, inputs, states):
         #     AB = sy.Matrix([[-2/tau, -1/tau**2, 1/tau**2], [1, 0, 0], [0, 0, 0]])
         #     dAB = sy.exp(dt * AB).simplify()
         #     dA, dB = dAB[:2, :2], dAB[:2, 2:]
-        tau = self.tau_var[:, None, None] + 1e-8
+        tau = tf.maximum(self.tau[:, None, None], 1e-8)
```
Comment (hunse) [inline keras_spiking/layers.py:909]: Not sure if we care about supporting float16, but the smallest value that float16 has is 6e-8. We could have an internal `eps` variable that depends on the layer's dtype.

Comment (hunse) [reply keras_spiking/layers.py:909]: I'm not sure if we test with float16 at all, though, so maybe we'll need a larger task for supporting float16.

Comment (drasmuss) [reply keras_spiking/layers.py:909]: Sorry merged before I saw this. But yeah I think supporting float16 would be a different (but worthwhile) enough thing to think about, so I'll make a new story for that.
